src/control/cmd/daos_server/storage_legacy.go

```diff
@@ -0,0 +1,216 @@
+	"github.com/daos-stack/daos/src/control/server/config"
+	"github.com/daos-stack/daos/src/control/server/storage"
+	"github.com/pkg/errors"
+)
+
+type legacyStorageCmd struct {
+	Prepare legacyPrepCmd `command:"prepare" description:"Prepare SCM and NVMe storage attached to local servers (deprecated, use scm (create|remove)-namespaces or nvme (prepare|release)-drives instead)."`
+	Scan    legacyScanCmd `command:"scan" description:"Scan SCM and NVMe storage attached to local server (deprecated, use scm scan or nvme scan instead)."`
+}
+
+type legacyPrepCmd struct {
+	cmdutil.LogCmd
+	scs   *server.StorageControlService
+	Reset bool `long:"reset" description:"Reset SCM modules to memory mode after removing namespaces. Reset SPDK returning NVMe device bindings back to kernel modules."`
+
+	ScmOnly               bool `short:"s" long:"scm-only" description:"Only prepare SCM."`
+	NrNamespacesPerSocket uint `short:"S" long:"scm-ns-per-socket" description:"Number of SCM namespaces to create per socket" default:"1"`
+	Force                 bool `short:"f" long:"force" description:"Perform format without prompting for confirmation"`
+
+	NvmeOnly     bool   `short:"n" long:"nvme-only" description:"Only prepare NVMe storage."`
+	PCIAllowList string `long:"pci-allow-list" description:"Whitespace separated list of PCI devices (by address) to be unbound from Kernel driver and used with SPDK (default is all PCI devices)."`
+	PCIBlockList string `long:"pci-block-list" description:"Whitespace separated list of PCI devices (by address) to be ignored when unbinding devices from Kernel driver to be used with SPDK (default is no PCI devices)."`
+	NrHugepages  int    `short:"p" long:"hugepages" description:"Number of hugepages to allocate for use by SPDK (default 1024)"`
+	TargetUser   string `short:"u" long:"target-user" description:"User that will own hugepage mountpoint directory and vfio groups."`
+	DisableVFIO  bool   `long:"disable-vfio" description:"Force SPDK to use the UIO driver for NVMe device access"`
+}
+
+// Validate checks both only options are not set and returns flags to direct
+// which subsystem types to prepare.
+func (cmd *legacyPrepCmd) Validate() (bool, bool, error) {
+	prepScm := cmd.ScmOnly || !cmd.NvmeOnly
+	if prepScm {
+		cmd.Debug("legacy prepare of scm selected")
+	}
+	prepNvme := cmd.NvmeOnly || !cmd.ScmOnly
+	if prepNvme {
+		cmd.Debug("legacy prepare of nvme selected")
+	}
+
+	if cmd.ScmOnly && cmd.NvmeOnly {
+		return false, false, errors.New(
+			"scm-only and nvme-only options should not be set together")
+	}
+
+	return prepScm, prepNvme, nil
+}
+
+func (cmd *legacyPrepCmd) prep(scs *server.StorageControlService, iommuEnabled bool) error {
+	var errSCM, errNVMe error
+
+	doSCM, doNVMe, err := cmd.Validate()
+	if err != nil {
+		return err
+	}
+
+	if cmd.Reset {
+		if doSCM {
+			var rnc removeNamespacesCmd
+			if err := convert.Types(cmd, &rnc); err != nil {
+				return errors.Wrap(err, "converting legacy prepare command")
+			}
```

Comment:
This makes me a little nervous that it could lead to subtle bugs if the fields don't get properly converted. I know this is legacy code now, but we have to support it for at least one release. Seems like it would be safer to just assign fields manually or else refactor a bit more so that the logic is callable directly from here.

Reply:
I can do but the coverage in storage_legacy_test.go is pretty comprehensive as it covers setting all the parameters in both the scm and nvme variants by checking the backend call request fields: https://github.com/daos-stack/daos/pull/9541/files#diff-23b564583a109ce64bcdbb49e4244db3aedbd2b0650b766c9b534ce199021c8bR85